### ISSUE TITLE
SC2154 false positive for cat EOF

### ISSUE THREAD

**Issue by matthewkayne** [koalaman/shellcheck#2481]
#### For bugs
- Rule Id (if any, e.g. SC1000): SC2154
- My shellcheck version (`shellcheck --version` or "online"): latest
- [x] The rule's wiki page does not already cover this (e.g. https://shellcheck.net/wiki/SC2086)
- [x] I tried on https://www.shellcheck.net/ and verified that this is still a problem on the latest commit

#### Here's a snippet or screenshot that shows the problem:

```sh
#!/bin/bash

cat > .gitignore << EOF
# System files
.DS_Store
Thumbs.db

# Visual Studio Code
.vscode

# Byte-compiled / optimized / DLL files
__pycache__/
*.py[cod]
*$py.class
EOF

```
<img width="296" alt="Screenshot 2022-04-04 at 16 36 39" src="https://user-images.githubusercontent.com/45180131/161580846-fd6c0241-398b-414e-834c-be8f180dd516.png">

#### Here's what shellcheck currently says:
In pyproject.sh line 14:
*$py.class
 ^-^ SC2[15](https://github.com/matthewkayne/Python-Project-Maker/runs/5819240377?check_suite_focus=true#step:4:15)4: py is referenced but not assigned.

For more information:
  https://www.shellcheck.net/wiki/SC2154 -- py is referenced but not assigned.
 ERROR: ShellCheck detected issues in pyproject.sh 🐛.

#### Here's what I wanted or expected to see:
It should have no error as `*$py.class` is within the EOF for the cat statement and is therefore part of a string, so this error should not be flagged.

**Comment by brother** [comment]
Either escape the dollar sign or enclose the EOF in single quotes to avoid variable expansion.
Slim example of start, shellcheck complains about this.
```shell
brother ~$ cat /tmp/test.sh
#!/bin/sh
cat <<EOF
string
$another
EOF
brother ~$ /tmp/test.sh 
string

```

Using single quotes.
```shell
brother ~$ cat /tmp/test.sh
#!/bin/sh
cat <<'EOF'
string
$another
EOF
brother ~$ /tmp/test.sh 
string
$another
```

Or escaping the "variable"
```shell
brother ~$ cat /tmp/test.sh
#!/bin/sh
cat <<EOF
string
\$another
EOF
brother ~$ /tmp/test.sh 
string
$another
```

**Comment by ab** [comment]
Just to expand on brother's reponse: this is not a false positive. Variables within heredocs are expanded unless escaped or the whole heredoc is single quoted.

**Comment by brother** [comment]
18 months later we can close this me thinks.
